crates/shim-sgx/src/handler/mod.rs

```diff
@@ -845,4 +879,78 @@ impl<'a> Handler<'a> {
             }
         }
     }
+
+    fn thread_mem_alloc(&mut self) -> sallyport::Result<*const Tcs> {
+        let usermemscope = UserMemScope;
+
+        let addr = self.mmap(
+            &usermemscope,
+            None,
+            size_of::<ThreadMem>(),
+            PROT_READ | PROT_WRITE,
+            MAP_PRIVATE | MAP_ANONYMOUS,
+            -1,
+            0,
+        )?;
+
+        // # Safety
+        // ThreadMem is a POD type and the memory is aligned and zeroed.
+        let tm = unsafe { &mut *(addr.as_ptr() as *mut ThreadMem) };
+
+        let stack_end_1 = NonNull::new(&mut tm.cssa_stack as *mut _ as *mut c_void).unwrap();
+        let stack_end_0 = NonNull::new(&mut tm.stack as *mut _ as *mut c_void).unwrap();
+        let tcs = &tm.tcs as *const Tcs;
+        let ssa = tm.ssa.as_ptr();
+
+        tm.tcs.ossa = ssa as u64 - shim_address() as u64;
```

Comment:
Can we get some explanatory comments throughout this function to help the reader follow along?

Reply:
yes, will do

Reply:
done